File: 2021/python/day2.py
DIRS = {
    "down" : (1, 0),
    "forward" : (0, 1),
    "up": (-1, 0)
}

def parse_line(line: str):
    dir, n = line.split()
    dx, dy = DIRS[dir]
    n = int(n)
    return n * dx, n * dy

def sim_part2(iter):
    x, y, aim = 0, 0, 0
    for action, X in iter:
        if action == 0: # then it was a `forward`
            x += X
            y += X * aim
        else:
            aim += action
    return x * y


def vector_add(iter):
    x, y = 0, 0
    for dx, dy in iter:
        x += dx
        y += dy
    return x, y

def part1(input):
    x, y = vector_add(map(parse_line, input))
    return x * y

if __name__ == "__main__":
    with open("inputs/day2.txt") as input:
        input = [*input]
        print(f"Part 1: {part1(input)}")
        print(f"Part 2: {sim_part2(map(parse_line, input))}")

File: 2021/rust/day2.rs
use std::io::{self, BufRead, BufReader};

#[derive(Debug, Clone, Copy)]
enum Action {
    Forward,
    Up,
    Down,
}

impl Action {
    fn parse(line: &str) -> Option<Self> {
        let it = match line {
            "forward" => Self::Forward,
            "up" => Self::Up,
            "down" => Self::Down,
            _ => return None,
        };
        Some(it)
    }
}


fn parse_line(line: &str) -> (Action, i32) {
    let mut it = line.split(' ');
    let action = it.next().and_then(Action::parse).unwrap();
    let num = it.next().unwrap().parse().unwrap();

    (action, num)
}

fn simulate_part1(actions: impl IntoIterator<Item = (Action, i32)>) -> i32 {
    let mut x = 0;
    let mut y = 0;
    for (action, dv) in actions {
        match action {
            Action::Forward => x += dv,
            Action::Up => y -= dv,
            Action::Down => y += dv,
        }
    }
    x * y
}

fn simulate_part2(actions: impl IntoIterator<Item = (Action, i32)>) -> i32 {
    let mut x = 0;
    let mut y = 0;
    let mut aim = 0;
    for (action, value) in actions {
        match action {
            Action::Forward => {
                x += value;
                y += value * aim;
            }
            Action::Up => {
                aim -= value;
            }
            Action::Down => aim += value,
        }
    }
    x * y
}



fn main() -> io::Result<()> {
    let actions: Vec<_> = std::fs::File::open("inputs/day2.txt").map(BufReader::new).map(BufReader::lines).and_then(|lines| {
        lines.map(|line| {
            let line = line?;
            Ok(parse_line(&line))
        }).collect::<Result<_, _>>()
    })?;
    println!("Part 1: {}", simulate_part1(actions.iter().cloned()));
    println!("Part 2: {}", simulate_part2(actions));
    Ok(())
}
    

